File: app/src/main/java/com/manrique/trailerstock/dao/VentaDao.java
package com.manrique.trailerstock.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import com.manrique.trailerstock.model.Venta;
import java.util.List;

@Dao
public interface VentaDao {

        @Insert
        long insertar(Venta venta); // Retorna el ID generado para usarlo en el detalle

        @Delete
        void eliminar(Venta venta);

        @Query("SELECT * FROM ventas ORDER BY fecha DESC")
        LiveData<List<Venta>> getAllVentas();

        @Query("SELECT * FROM ventas ORDER BY fecha DESC")
        List<Venta> obtenerTodas();

        // Consulta para saber cuanto se vendio hoy
        @Query("SELECT SUM(total) FROM ventas WHERE fecha >= :inicioDia")
        double obtenerTotalVendidoDia(long inicioDia);

        @Query("SELECT * FROM ventas WHERE id = :id")
        Venta obtenerPorId(int id);

        @Query("SELECT * FROM ventas WHERE aplicoPromo = 1 ORDER BY fecha DESC")
        List<Venta> obtenerVentasConPromocion();

        // ===== QUERIES PARA ESTADÍSTICAS =====

        // Ventas del día
        @Query("SELECT SUM(total) FROM ventas WHERE fecha >= :inicioDelDia")
        Double obtenerTotalVentasDelDia(long inicioDelDia);

        @Query("SELECT COUNT(*) FROM ventas WHERE fecha >= :inicioDelDia")
        int obtenerCantidadVentasDelDia(long inicioDelDia);

        // Ventas del mes
        @Query("SELECT SUM(total) FROM ventas WHERE fecha >= :inicioDelMes")
        Double obtenerTotalVentasDelMes(long inicioDelMes);

        @Query("SELECT COUNT(*) FROM ventas WHERE fecha >= :inicioDelMes")
        int obtenerCantidadVentasDelMes(long inicioDelMes);

        // Ventas por tipo de cliente
        @Query("SELECT SUM(total) FROM ventas WHERE tipoCliente = :tipo AND fecha >= :desde")
        Double obtenerVentasPorTipo(String tipo, long desde);

        // Ventas con promociones
        @Query("SELECT COUNT(*) FROM ventas WHERE aplicoPromo = 1 AND fecha >= :desde")
        int obtenerCantidadVentasConPromocion(long desde);

        // Promedio de venta
        @Query("SELECT AVG(total) FROM ventas WHERE fecha >= :desde")
        Double obtenerPromedioVentas(long desde);

        // ===== NUEVAS QUERIES PARA MEJORAS DE ESTADÍSTICAS =====

        // Obtener ventas agrupadas por día para gráficos
        @Query("SELECT strftime('%Y-%m-%d', fecha/1000, 'unixepoch', 'localtime') as fecha, " +
                        "SUM(total) as totalVentas, " +
                        "COUNT(*) as cantidadVentas " +
                        "FROM ventas " +
                        "WHERE fecha >= :desde AND fecha <= :hasta " +
                        "GROUP BY strftime('%Y-%m-%d', fecha/1000, 'unixepoch', 'localtime') " +
                        "ORDER BY fecha ASC")
        List<com.manrique.trailerstock.model.VentasPorDia> obtenerVentasPorDia(long desde, long hasta);

        // Obtener listado de ventas de un día específico con información resumida
        @Query("SELECT v.id as ventaId, v.fecha, v.total, v.tipoCliente, v.aplicoPromo, " +
                        "COUNT(vd.id) as cantidadProductos " +
                        "FROM ventas v " +
                        "LEFT JOIN venta_detalles vd ON v.id = vd.ventaId " +
                        "WHERE v.fecha >= :inicioDia AND v.fecha < :finDia " +
                        "GROUP BY v.id " +
                        "ORDER BY v.fecha DESC")
        List<com.manrique.trailerstock.model.VentaConDetalles> obtenerVentasDelDia(long inicioDia, long finDia);

        // Obtener ventas en un rango de fechas
        @Query("SELECT * FROM ventas WHERE fecha >= :desde AND fecha <= :hasta ORDER BY fecha DESC")
        List<Venta> obtenerVentasPorRango(long desde, long hasta);

        // Calcular total de costos (necesario para margen de ganancia)
        @Query("SELECT SUM(vd.cantidad * p.precio_costo) " +
                        "FROM venta_detalles vd " +
                        "INNER JOIN productos p ON vd.productoId = p.id " +
                        "INNER JOIN ventas v ON vd.ventaId = v.id " +
                        "WHERE v.fecha >= :desde")
        Double obtenerCostoTotalVentas(long desde);
}

File: app/src/main/java/com/manrique/trailerstock/data/local/dao/VentaDao.kt
package com.manrique.trailerstock.data.local.dao

import androidx.room.*
import com.manrique.trailerstock.data.local.entities.Venta
import kotlinx.coroutines.flow.Flow

/**
 * DAO para operaciones de Venta con soporte de Coroutines.
 */
@Dao
interface VentaDao {
    
    @Query("SELECT * FROM ventas ORDER BY fecha DESC")
    fun obtenerTodas(): Flow<List<Venta>>
    
    @Query("SELECT * FROM ventas WHERE id = :id")
    suspend fun obtenerPorId(id: Int): Venta?
    
    @Insert
    suspend fun insertar(venta: Venta): Long
    
    @Update
    suspend fun actualizar(venta: Venta)
    
    @Delete
    suspend fun eliminar(venta: Venta)
    
    // Ventas del día
    @Query("SELECT * FROM ventas WHERE fecha >= :timestampInicio AND fecha < :timestampFin ORDER BY fecha DESC")
    fun obtenerVentasPorFecha(timestampInicio: Long, timestampFin: Long): Flow<List<Venta>>
    
    // Total de ventas del día
    @Query("SELECT SUM(total) FROM ventas WHERE fecha >= :timestampInicio AND fecha < :timestampFin AND estado = 'ACTIVA'")
    suspend fun obtenerTotalPorFecha(timestampInicio: Long, timestampFin: Long): Double?
    
    // Total del mes
    @Query("SELECT SUM(total) FROM ventas WHERE fecha >= :timestampInicio AND estado = 'ACTIVA'")
    suspend fun obtenerTotalDesde(timestampInicio: Long): Double?
    
    // Ventas por tipo de cliente
    @Query("SELECT SUM(total) FROM ventas WHERE tipo_cliente = :tipoCliente AND fecha >= :timestampInicio AND estado = 'ACTIVA'")
    suspend fun obtenerTotalPorTipoCliente(tipoCliente: String, timestampInicio: Long): Double?
    
    // Ticket promedio
    @Query("SELECT AVG(total) FROM ventas WHERE fecha >= :timestampInicio AND estado = 'ACTIVA'")
    suspend fun obtenerTicketPromedio(timestampInicio: Long): Double?
    
    // Cantidad de ventas
    @Query("SELECT COUNT(*) FROM ventas WHERE fecha >= :timestampInicio AND fecha < :timestampFin AND estado = 'ACTIVA'")
    suspend fun contarVentasPorFecha(timestampInicio: Long, timestampFin: Long): Int
    
    // Filtros para búsqueda de ventas
    @Query("SELECT * FROM ventas WHERE metodo_pago = :metodoPago ORDER BY fecha DESC")
    fun obtenerPorMetodoPago(metodoPago: String): Flow<List<Venta>>
    
    @Query("SELECT * FROM ventas WHERE tipo_cliente = :tipoCliente ORDER BY fecha DESC")
    fun obtenerPorTipoCliente(tipoCliente: String): Flow<List<Venta>>
    
    @Query("SELECT * FROM ventas WHERE fecha >= :fechaInicio AND fecha <= :fechaFin ORDER BY fecha DESC")
    fun obtenerPorRangoFechas(fechaInicio: Long, fechaFin: Long): Flow<List<Venta>>

    // Ganancia estimada (Ventas - Costo)
    @Query("""
        SELECT SUM(vd.subtotal - (vd.cantidad * p.precio_costo)) 
        FROM venta_detalles vd 
        JOIN productos p ON vd.producto_id = p.id 
        JOIN ventas v ON vd.venta_id = v.id 
        WHERE v.fecha >= :timestampInicio AND v.estado = 'ACTIVA'
    """)
    suspend fun obtenerGananciaEstimada(timestampInicio: Long): Double?

    // Top productos más vendidos
    @Query("""
        SELECT p.nombre as nombre, SUM(vd.cantidad) as cantidadVendida 
        FROM venta_detalles vd 
        JOIN productos p ON vd.producto_id = p.id 
        JOIN ventas v ON vd.venta_id = v.id 
        WHERE v.fecha >= :timestampInicio AND v.estado = 'ACTIVA' 
        GROUP BY vd.producto_id 
        ORDER BY cantidadVendida DESC 
        LIMIT :limit
    """)
    suspend fun obtenerTopProductos(timestampInicio: Long, limit: Int): List<ProductoVendido>
}

/**
 * Pojo para los productos más vendidos
 */
data class ProductoVendido(
    val nombre: String,
    val cantidadVendida: Int
)
